File: cpp/dnn_control/fullstatecontroller.cpp
#include "fullstatecontroller.h"

FullStateController::FullStateController(const double &control_frequency, const Vector3D &target_position) : SpacecraftController(7),
    control_interval_(1.0 / control_frequency) {
    for (int i = 0; i < 3; ++i) {
        target_position_[i] = target_position[i];
        previous_error_[i] = 0.0;
        integral_[i] = 0.0;
    }

    constant_integral_ = 0.0;
    constant_derivative_ = 5;
    constant_proportional_ = 4;
}

FullStateController::~FullStateController() {

}

void FullStateController::GetThrustForSensorData(const SensorData &sensor_data, Vector3D &thrust) {
    for (int i = 0; i < 3; ++i) {
        const double error = target_position_[i] - sensor_data[i];
        const double derivative = (error - previous_error_[i]) / control_interval_;
        integral_[i] += error * control_interval_;
        thrust[i] = 0.0; //constant_proportional_ * error + constant_derivative_ * derivative + constant_integral_ * integral_[i];
        previous_error_[i] = error;
    }
}

File: python/dnn_control/fullstatecontroller.py
class FullStateController:

    def __init__(self, target_position, control_interval):
        self._target_position = target_position
        self._control_interval = control_interval
        self._previous_error = [0.0, 0.0, 0.0]
        self._integral = [0.0, 0.0, 0.0]
        self._constant_proportional = 1.0
        self._constant_derivative = 1.5
        self._constant_integral = 0.0

    def get_thrust_for_sensor_data(self, sensor_data):
        current_position = sensor_data[0:3]
        thrust = [0.0, 0.0, 0.0]
        for i in [0, 1, 2]:
            error = self._target_position[i] - current_position[i]
            self._integral[i] += error * self._control_interval
            derivative = (error - self._previous_error[i]) / self._control_interval
            thrust[i] = self._constant_proportional * error\
                        + self._constant_derivative * derivative \
                        + self._constant_integral * self._integral[i]
            self._previous_error[i] = error

        return thrust
